### ISSUE TITLE
HEALPix curriculum

### ISSUE THREAD

**Issue by TillHae** [ecmwf/WeatherGenerator#2794]
### Describe the task. It can be a feature, documentation, etc.

Add HEALPix curriculum to the WeatherGenerator for more efficient training.

In my bachelor thesis, I discovered that starting at a lower HEALPix Level (HL) and sequentially increasing it during training results in efficiency gains. This coarse-to-fine curriculum reduces compute time by almost 95% in the early stages of training up to HL 6. Furthermore, it continues to save compute to a smaller extent as the curriculum progresses, all without losing final model performance. When training to even finer resolutions like HL 8 it should even result in a greater efficiency gain. We should integrate a configuration mechanism to automate these HL transitions and dataset swaps natively during training.

### Area

- [ ] datasets, data readers, data preparation and transfer
- [x] model
- [ ] science
- [ ] infrastructure and engineering
- [ ] evaluation, export and visualization
- [ ] documentation
- [ ] performance

**Comment by TillHae** [comment]
Assign Me

**Comment by clessig** [comment]
@TillHae : do you have any quantitative results for how updating the healpix level increases training efficiency? Is this only at the input and output or also in latent space?

**Comment by TillHae** [comment]
The latent space isn't affected when updating the HEALPix level. On Monday, I presented at the High-Res meeting, and they asked me to run training on more samples to see how it impacts generalization. Regarding efficiency, it improves a lot, as you can see in the following plot. The black line is the baseline, which was run on HEALPix level 6 for the entire time.

<img width="2952" height="1752" alt="Image" src="https://github.com/user-attachments/assets/276b9636-771c-4801-ba21-6bc81c81bfec" />

**Comment by clessig** [comment]
> The latent space isn't affected when updating the HEALPix level. On Monday, I presented at the High-Res meeting, and they asked me to run training on more samples to see how it impacts generalization. Regarding efficiency, it improves a lot, as you can see in the following plot. The black line is the baseline, which was run on HEALPix level 6 for the entire time.
> 
> <img alt="Image" width="2000" height="1187" src="https://private-user-images.githubusercontent.com/173140033/646959415-276b9636-771c-4801-ba21-6bc81c81bfec.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODg3MjMzMDEsIm5iZiI6MTc4ODcyMzAwMSwicGF0aCI6Ii8xNzMxNDAwMzMvNjQ2OTU5NDE1LTI3NmI5NjM2LTc3MWMtNDgwMS1iYTIxLTZiYzgxYzgxYmZlYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwOTA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDkwNlQxOTMwMDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zNTE4MDBkZjczMmMyOWM4NTA1NTY3MWVmZDc5MWY1MTMyNjQ5NDJlYzEzNmYwZDJiNWNkOWYzMDgyZDcwNTIyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZyZXNwb25zZS1jb250ZW50LXR5cGU9aW1hZ2UlMkZwbmcifQ.0TJXAWgN6QpSzdKAORbX5RAOv-fRtwdUAyKizJ6MQ84">

Could you share the loss curves for the standard headline variables.

**Comment by clessig** [comment]
In latent space, the healpix level is always 6 in these experiments and it only changes for embedding and decoding?

**Comment by TillHae** [comment]
> In latent space, the healpix level is always 6 in these experiments and it only changes for embedding and decoding?

Sorry, I mixed things up. The latent space does change by a factor of 4 but the model size doesn't. The only thing where we encounter a problem is with the attention mechanism. After a transition the old attention map doesn't make any sense and this is the loss spike that is then visible in the plot.

**Comment by TillHae** [comment]
> Could you share the loss curves for the standard headline variables.

I didn't even know that it was possible to plot the training loss by variable. Which script do I need to run to get those results?

**Comment by clessig** [comment]
> > Could you share the loss curves for the standard headline variables.
> 
> I didn't even know that it was possible to plot the training loss by variable. Which script do I need to run to get those results?

`uv run plot_train --channels z_500 u_850 v_850 t_850 q_850 2t 10u 10v`

**Comment by clessig** [comment]
> > In latent space, the healpix level is always 6 in these experiments and it only changes for embedding and decoding?
> 
> Sorry, I mixed things up. The latent space does change by a factor of 4 but the model size doesn't. The only thing where we encounter a problem is with the attention mechanism. After a transition the old attention map doesn't make any sense and this is the loss spike that is then visible in the plot.

The attention map (you mean attention values, i.e. normalized softmax of the score matrix?) is determined by the input tokens and not learned. Do you mean something else?

**Comment by TillHae** [comment]
> > > In latent space, the healpix level is always 6 in these experiments and it only changes for embedding and decoding?
> > 
> > 
> > Sorry, I mixed things up. The latent space does change by a factor of 4 but the model size doesn't. The only thing where we encounter a problem is with the attention mechanism. After a transition the old attention map doesn't make any sense and this is the loss spike that is then visible in the plot.
> 
> The attention map (you mean attention values, i.e. normalized softmax of the score matrix?) is determined by the input tokens and not learned. Do you mean something else?

What I meant is that the query and key vectors generated when the input tokens are applied to the **learned weights of the global AE** make no sense and therefore the "attention map" should also make no sense.

Here is the calculation: [query and key calculation in ll. 595-600 of the attention](https://github.com/ecmwf/WeatherGenerator/blob/develop/src/weathergen/model/attention.py#L595-L600)

I think what you mean is [ll. 953-975 of the PyTorch attention](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/transformers/attention.cpp#L953-L975), which we use. But as I just pointed out the input for this calculation should be impacted by the learned weights of the global AE, which results in the output of the calculation being basically garbage at the beginning of a HL transition. But tbh I don't have a full overview of what is actually going on throughout the entire training process.

I think in the process of looking at everything I also found dead code:
`self.att` in [ll. 583-588 of the attention](https://github.com/ecmwf/WeatherGenerator/blob/develop/src/weathergen/model/attention.py#L583-L588) is actually never used in the forward call of the `MultiSelfAttentionHead`.

**Comment by TillHae** [comment]
> > > Could you share the loss curves for the standard headline variables.
> > 
> > 
> > I didn't even know that it was possible to plot the training loss by variable. Which script do I need to run to get those results?
> 
> `uv run plot_train --channels z_500 u_850 v_850 t_850 q_850 2t 10u 10v`

[Hedgedoc with plots](https://gitlab.jsc.fz-juelich.de/hedgedoc/VSkaXd6_QQSRluxj6UMYVQ?view)

**Comment by clessig** [comment]
> > > > Could you share the loss curves for the standard headline variables.
> > > 
> > > 
> > > I didn't even know that it was possible to plot the training loss by variable. Which script do I need to run to get those results?
> > 
> > 
> > `uv run plot_train --channels z_500 u_850 v_850 t_850 q_850 2t 10u 10v`
> 
> [Hedgedoc with plots](https://gitlab.jsc.fz-juelich.de/hedgedoc/VSkaXd6_QQSRluxj6UMYVQ?view)

The improvements you show as a function of training time are tantilizing but more complex if one looks at the loss as a function of samples, with the number of samples being very low (we usually train for at least 600,000 steps for a good model).

Could you train for 500,000 steps and plot again loss as a function of wall clock and samples. (It would be great to have the wall clock option also in plot_train).

**Comment by TillHae** [comment]
> > [Hedgedoc with plots](https://gitlab.jsc.fz-juelich.de/hedgedoc/VSkaXd6_QQSRluxj6UMYVQ?view)
> 
> The improvements you show as a function of training time are tantilizing but more complex if one looks at the loss as a function of samples, with the number of samples being very low (we usually train for at least 600,000 steps for a good model).
> 
> Could you train for 500,000 steps and plot again loss as a function of wall clock and samples. (It would be great to have the wall clock option also in plot_train).

Here are some results @Tewson1 got when training for over 1 million samples from HL5 to HL8: 
https://gitlab.jsc.fz-juelich.de/hedgedoc/0HjLMt32RB6L2ZFY42msFw?view

As you can see in the final wall-clock plot, the Curriculum strategy remains ~5 hours ahead in validation loss. The relative time saved is smaller than in my thesis because this run uses a different global AE, which eliminates the central bottleneck.

Throughput comparison (samples/sec):
**@Tewson1's longer run:**
* HL5: 2.55 s/sec
* HL6: 2.26 s/sec
* HL7: 2.08 s/sec
* HL8: 1.51 s/sec

**My thesis run:**
* HL3: 1.75 s/sec
* HL4: 1.65 s/sec
* HL5: 0.64 s/sec
* HL6: 0.12 s/sec

I guess it is expectable to see no benefit in terms of compute when their is no big throughput advantage at lower HL
